pkg/vault/test/foundry/E2eSwap.t.sol

```diff
@@ -0,0 +1,696 @@
+        vm.stopPrank();
+    }
+
+    /**
+     * @notice Set up tokens.
+     * @dev When extending the test, override this function and set the same variables.
+     */
+    function setUpTokens() internal virtual {
+        tokenA = dai;
+        tokenB = usdc;
+    }
+
+    /**
+     * @notice Set up test variables (sender, poolCreator, pool swap fee, swap sizes).
+     * @dev When extending the test, override this function and set the same variables.
+     */
+    function setUpVariables() internal virtual {
+        sender = lp;
+        poolCreator = lp;
+
+        // If there are swap fees, the amountCalculated may be lower than MIN_TRADE_AMOUNT. So, multiplying
+        // MIN_TRADE_AMOUNT by 10 creates a margin.
+        minSwapAmountTokenA = 10 * MIN_TRADE_AMOUNT;
+        maxSwapAmountTokenA = poolInitAmountTokenA;
+
+        minSwapAmountTokenB = 10 * MIN_TRADE_AMOUNT;
+        maxSwapAmountTokenB = poolInitAmountTokenB;
+
+        // 0.0001% min swap fee.
+        minPoolSwapFeePercentage = 1e12;
+        // 10% max swap fee.
+        maxPoolSwapFeePercentage = 10e16;
+    }
+
+    /// @notice Donate tokens to vault, so liquidity tests are possible.
+    function _donateToVault() internal virtual {
+        tokenA.mint(address(vault), 100 * poolInitAmountTokenA);
+        tokenB.mint(address(vault), 100 * poolInitAmountTokenB);
+        // Override vault liquidity, to make sure the extra liquidity is registered.
+        vault.manualSetReservesOf(tokenA, 100 * poolInitAmountTokenA);
+        vault.manualSetReservesOf(tokenB, 100 * poolInitAmountTokenB);
+    }
+
+    function testDoUndoExactInNoFees__Fuzz(uint256 exactAmountIn) public {
+        exactAmountIn = bound(exactAmountIn, minSwapAmountTokenA, maxSwapAmountTokenA);
+
+        // Set swap fees to 0 (do not check pool fee percentage limits; some pool types do not accept 0 fees).
+        vault.manualUnsafeSetStaticSwapFeePercentage(pool, 0);
+
+        BaseVaultTest.Balances memory balancesBefore = getBalances(sender);
+
+        vm.startPrank(sender);
+        uint256 exactAmountOutDo = router.swapSingleTokenExactIn(
+            pool,
+            tokenA,
+            tokenB,
+            exactAmountIn,
+            0,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+        uint256 exactAmountOutUndo = router.swapSingleTokenExactIn(
+            pool,
+            tokenB,
+            tokenA,
+            exactAmountOutDo,
+            0,
+            MAX_UINT128,
+            false,
+            bytes("")
+        );
+        vm.stopPrank();
+
+        BaseVaultTest.Balances memory balancesAfter = getBalances(sender);
+
+        assertLe(exactAmountOutUndo, exactAmountIn, "Amount out undo should be <= exactAmountIn");
+
+        _checkUserBalancesAndPoolInvariant(balancesBefore, balancesAfter, 0, 0);
+    }
+
+    function testDoUndoExactInLiquidity__Fuzz(uint256 liquidityTokenA, uint256 liquidityTokenB) public {
```

Comment:
I'd suggest a modification over this one: 
- Have an internal function that receives initial liquidity and swap sizes
- Fuzz initial liquidity, keep swap size the same (i.e. what this test does now)
- Keep initial liquidity fixed, fuzz swap size
- Fuzz all

(this can be done in a follow-up PR; you might need to add a few more bounds here and there).

Reply:
I think, with the latest version, this was addressed.